src/core/thread/mesh_forwarder.cpp

```diff
@@ -1415,15 +1415,18 @@ void MeshForwarder::HandleReceivedFrame(Mac::Frame &aFrame, ThreadError aError)
     switch (aFrame.GetType())
     {
     case Mac::Frame::kFcfFrameData:
-        if (reinterpret_cast<Lowpan::MeshHeader *>(payload)->IsMeshHeader())
+        if (payloadLength >= sizeof(Lowpan::MeshHeader) &&
+            reinterpret_cast<Lowpan::MeshHeader *>(payload)->IsMeshHeader())
         {
             HandleMesh(payload, payloadLength, messageInfo);
         }
-        else if (reinterpret_cast<Lowpan::FragmentHeader *>(payload)->IsFragmentHeader())
+        else if (payloadLength >= sizeof(Lowpan::FragmentHeader) &&
+                 reinterpret_cast<Lowpan::FragmentHeader *>(payload)->IsFragmentHeader())
         {
             HandleFragment(payload, payloadLength, macSource, macDest, messageInfo);
         }
-        else if (Lowpan::Lowpan::IsLowpanHc(payload))
+        else if (payloadLength >= 1 &&
+                 Lowpan::Lowpan::IsLowpanHc(payload))
         {
             HandleLowpanHC(payload, payloadLength, macSource, macDest, messageInfo);
         }
```

Comment:
I don't want to go through all the code paths right now, but someone needs to. If you go through the receive data paths, you will see that a lot of the code assumes the buffer is always available.

If we annotated all the functions with [SAL](https://msdn.microsoft.com/en-us/library/ms182032.aspx) (a lot of work, but might be worth it), the Windows build can catch these things at build time. It is a large change, and some would say it makes the functions look ugly (though you get use to it in time), it has major advantages. All Windows code uses it internally. You can see I already use it in my driver code.